src/node/sv2_template_provider.cpp

```diff
@@ -0,0 +1,404 @@
+}
+
+fs::path Sv2TemplateProvider::GetAuthorityKeyFile()
+{
+    return gArgs.GetDataDirNet() / "sv2_authority_key";
+}
+
+bool Sv2TemplateProvider::Start(const Sv2TemplateProviderOptions& options)
+{
+    std::string host = "127.0.0.1";
+    uint16_t port{options.port};
+    if (gArgs.IsArgSet("-sv2bind")) { // Specific bind address
+        std::optional<std::string> sv2_bind{gArgs.GetArg("-sv2bind")};
+        if (sv2_bind) {
+            if (!SplitHostPort(sv2_bind.value(), port, host)) {
+                throw std::runtime_error(strprintf("Invalid port %d", options.port));
+            }
+        }
+    }
+
+    m_minimum_fee_delta = gArgs.GetIntArg("-sv2feedelta", DEFAULT_SV2_FEE_DELTA);
+
+    if (!m_connman->Start(this, host, port)) {
+        return false;
+    }
+
+    m_thread_sv2_handler = std::thread(&util::TraceThread, "sv2", [this] { ThreadSv2Handler(); });
+    return true;
+}
+
+Sv2TemplateProvider::~Sv2TemplateProvider()
+{
+    AssertLockNotHeld(m_tp_mutex);
+
+    m_connman->Interrupt();
+    m_connman->StopThreads();
+
+    Interrupt();
+    StopThreads();
+}
+
+void Sv2TemplateProvider::Interrupt()
+{
+    m_flag_interrupt_sv2 = true;
+}
+
+void Sv2TemplateProvider::StopThreads()
+{
+    if (m_thread_sv2_handler.joinable()) {
+        m_thread_sv2_handler.join();
+    }
+}
+
+class Timer {
+private:
+    std::chrono::seconds m_interval;
+    std::chrono::seconds m_last_triggered;
+
+public:
+    Timer() {
+        m_interval = std::chrono::seconds(gArgs.GetIntArg("-sv2interval", DEFAULT_SV2_INTERVAL));
+        // Initialize the timer to a time point far in the past
+        m_last_triggered = GetTime<std::chrono::seconds>() - std::chrono::hours(1);
+    }
+
+    bool trigger() {
+        auto now{GetTime<std::chrono::seconds>()};
+        if (now - m_last_triggered >= m_interval) {
+            m_last_triggered = now;
+            return true;
+        }
+        return false;
+    }
+};
+
+void Sv2TemplateProvider::ThreadSv2Handler()
+{
+    Timer timer;
+    unsigned int mempool_last_update = 0;
+    unsigned int template_last_update = 0;
+
+    while (!m_flag_interrupt_sv2) {
+
+        bool best_block_changed = [this]() {
+            if (m_mining.isInitialBlockDownload() && gArgs.GetChainType() != ChainType::SIGNET) {
+                // Keep m_best_prev_hash unset during IBD to avoid pushing outdated
+                // templates. Except for signet, because we might be the only miner.
+                return false;
+            }
+
+            WAIT_LOCK(g_best_block_mutex, lock);
+            auto checktime = std::chrono::steady_clock::now() + std::chrono::milliseconds(50);
+            g_best_block_cv.wait_until(lock, checktime);
+            LOCK(m_tp_mutex);
+            if (m_best_prev_hash != g_best_block && g_best_block != uint256(0)) {
+                m_best_prev_hash = g_best_block;
+                return true;
+            }
+            return false;
+        }();
```

Comment:
How is this not polling?

Reply:
It's not polling over RPC.

Reply:
> How is this not polling?

I was reading it as "check every 50ms" but I just noticed that it is also using a condvar, so it'll get notified when `g_best_block` changes.